python/ray/dag/compiled_dag_node.py

```diff
@@ -386,6 +324,79 @@ def __init__(
             assert not isinstance(val, ChannelInterface)
             assert not isinstance(val, DAGInputAdapter)
 
+        self.input_reader: ReaderInterface = SynchronousReader(self.input_channels)
+        self.output_writer: WriterInterface = SynchronousWriter(self.output_channel)
+        self._cache = None
+
+    def cancel(self):
+        self.input_reader.close()
+        self.output_writer.close()
+
+    def prepare(self):
+        for typ_hint in self.input_type_hints:
+            typ_hint.register_custom_serializer()
+        self.output_type_hint.register_custom_serializer()
+        self.input_reader.start()
+        self.output_writer.start()
+
+    def set_cache(self, data):
+        self._cache = data
+
+    def reset_cache(self):
+        data = self._cache
+        self._cache = None
+        return data
+
+    def _read(self):
+        try:
+            res = self.input_reader.read()
+            self.set_cache(res)
+            return False
+        except RayChannelError:
+            # Channel closed. Exit the loop.
+            return True
+
+    def _compute(self, actor_handle):
+        res = self.reset_cache()
+        method = getattr(actor_handle, self.method_name)
+        try:
+            _process_return_vals(res, return_single_output=False)
+        except Exception as exc:
+            # Previous task raised an application-level exception.
+            # Propagate it and skip the actual task. We don't need to wrap the
+            # exception in a RayTaskError here because it has already been wrapped
+            # by the previous task.
+            self.set_cache(exc)
```

Comment:
Good to know. This is just copied from https://github.com/ray-project/ray/blob/b299b0bf2be6f9c685aec970b15677b588254aff/python/ray/dag/compiled_dag_node.py#L182

I will update it.

Reply:
It breaks some tests. How about I update this in a follow-up PR instead since this is just a copy and paste from the existing codebase so that this PR will be easier to review?

Reply:
Open an issue: https://github.com/ray-project/ray/issues/47010

Reply:
Let me know if you prefer to update in this PR.

Reply:
Resolve for now. Feel free to reopen if you prefer to update it in this PR.